Address issue #38: Make general arbitrable deposit contract

l-hartman opened the pull request:
Hello!
This is my attempt at issue #38. Let me know if my implementation is at all wrong.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -0,0 +1,80 @@
+        amount += msg.value;
+    }
+
+    /** @dev Owner deposit to contract. To be called when the owner
```
I think the end of the comment is missing.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -0,0 +1,80 @@
+
+    /** @dev Owner response to claimant. To be called when the owner initates a
+     *  a response to the claimant. 
+     *  @param 
```
We must fill the doxygen-style comment for _responseAmount

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -0,0 +1,80 @@
+
+     /** @dev Settles claim and pays Party B. Transfers the owner's deposit to the claimant 
+     */
+    function settleClaim() private{
```
settleClaim is only used once, so no need a specific function as it increases code complexity.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -0,0 +1,80 @@
+
+    function ArbitrableDeposit (
+        Arbitrator _arbitrator, bytes32 _hashContract, uint _timeout,
+        address _partyB, bytes _arbitratorExtraData)
```
In this case partyB is specified, so it's not 

> Any party may make a claim

We must allow anyone to make claims, so this contract will probably not inherit from TwoPartyArbitrable.

l-hartman commented:
I was unsure where to initialize claimRate 'r' as described in the issue. I decided to do so in the constructor. 

I have made sizable changes as generalized below.
- ArbitrableDeposit now inherits Arbitrable instead on TwoPartyArbitrable. Though, I did copy a few functions from TwoPartyArbitrable.
- Anyone who is not the owner can initiate a claim. 
- Added more functionalities for arbitration.
- Fixed comments. 

I am unsure how kleros' testing works, but it compiles!  Let me know if more changes are needed.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
- *  This is a a contract for a deposit by the owner where a claimant can seek arbitration. 
- *  Party A is the owner and Party B is the claimant.
+ *  This is a a contract which allow for an owner deposit. Anyone besides the owner can seek arbitration/file a claim as a claimant. 
+ * To develop a contract inheriting from this one, you need to:
```
This seems to come from the Arbitrable contract.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
-        address(this).transfer(amount);
+    function deposit(uint _amount) onlyOwner {
+        amount += _amount;
+        address(this).transfer(_amount);
```
It seems to transfer some funds to itself and is not payable.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
+     *  @param _claimAmount The proposed claim amount by the claimant.
+     */    
+    function makeClaim(uint _claimAmount) onlyNotOwner {
+        require(_claimAmount >= 0 && _claimAmount <= amount);
```
_claimAmount >= 0 is always verified since _claimAmount is a uint.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
+    /** @dev File a claim against owner. To be called when someone makes a claim.
+     *  @param _claimAmount The proposed claim amount by the claimant.
+     */    
+    function makeClaim(uint _claimAmount) onlyNotOwner {
```
What is the purpose of the onlyNotOwner?
Would it lead to a vulnerability if the owner himself make a claim?

clesaege replied:
This need to be payable to have a claimDeposit.
We must also ask the claimer to deposit arbitration fee upfront to avoid frivolous claims.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
+     *  @param _responseAmount The counter-offer amount the Owner proposes to a claimant.
+     */ 
+    function claimResponse(uint _responseAmount) onlyOwner { 
+        require(_responseAmount >= 0 && _responseAmount <= claimDepositAmount);
```
_responseAmount >= 0 is always true since _responseAmount is a uint

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
+        status = Status.NoDispute;
         amount += msg.value;
+        owner = msg.sender;
+        address(this).transfer(amount);
```
What is the point of transferring amount to the contract?

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
+        require(_claimAmount >= 0 && _claimAmount <= amount);
+        claimant = msg.sender;
+        claimAmount = _claimAmount;
+        claimDepositAmount = (_claimAmount * claimRate)/100;
```
Can we put 100 in a constant? This is to avoid magic numbers.

clesaege commented on contracts/standard/arbitration/ArbitrableDeposit.sol:
```diff
@@ -7,74 +6,178 @@
+        } else if (_ruling==CLAIMANT_WINS)
+            claimant.transfer(amount);
         amount = 0;
-    }
```
Is there a way for the owner to withdraw the ETH?
